### PR TITLE
Clean up junk & add caching to stability

### PR DESCRIPTION
**Pull request by arielb1**
The caching essentially eliminates "stability checking" time (my attempt to clean-up junk got tangled up with stability, so I added the caching while I was at it).

r? @eddyb

### CODE REVIEW

File: src/librustc_typeck/coherence/mod.rs [hunk 1]
```diff
@@ -287,11 +275,8 @@ impl<'a, 'tcx> CoherenceChecker<'a, 'tcx> {
                     }
                 }).collect();
 
-                if opt_trait.is_some() {
-                    let trait_ref = ty::impl_id_to_trait_ref(self.crate_context.tcx,
-                                                             item.id);
-
-                    self.instantiate_default_methods(local_def(item.id),
+                if let Some(trait_ref) = ty::impl_trait_ref(self.crate_context.tcx,
+                                                            local_def(item.id)) {                    self.instantiate_default_methods(local_def(item.id),
```
Comment (eddyb) [inline src/librustc_typeck/coherence/mod.rs]: Looks like the newline got eaten (found by Travis).

**Comment by eddyb** [comment]
This is very nice cleanup, thanks! r=me with nit fixed

**Comment by arielb1** [comment]
@eddyb: fixed

**Comment by arielb1** [comment]
For real this time

**Pull request merged**